### PR TITLE
Set IPv6/local-ip to the node.

### PR DESCRIPTION
**Pull request by sergelogvinov**
I am using the self-hosted Kubernetes.
And the private network is used for node-to-node communications.

I've added the local network and IPv6 support to the CCM.

### CODE REVIEW

File: exoscale/sks_agent_runner_node_csr_validation.go [hunk 1]
```diff
@@ -104,30 +106,42 @@ func (r *sksAgentRunnerNodeCSRValidation) run(ctx context.Context) {
...
+
+						sort.Strings(addrs)
+						sort.Strings(addrsCSR)
+
+						if !reflect.DeepEqual(addrs, addrsCSR) {
+							errorf("sks-agent: CSR %s Node IP address doesn't match corresponding "+
+								"Compute instance public interface IP address, %q, got %q", csr.Name, addrs, parsedCSR.IPAddresses)
+
 							break
 						}
```
Comment (PhilippeChepy) [inline exoscale/sks_agent_runner_node_csr_validation.go:139]: The CSR object contains ip addresses from the kubelet configuration (e.g. Kubelet `--node-ip` argument).

When not specified it defaults to eth0 ipv4 IIRC.
But other valid use cases can also be:
- only the public IPv6
- only a single private IP from a privnet
- all ip addresses assigned to the instance (what this PR handles).

In order to handle all use cases, `addrsCSR` must be a subset of `addrs`, not necessarily equal

**Review by PhilippeChepy — COMMENTED** [review]
Hello @sergelogvinov,

Thank you for your PR! This is a very interesting contribution.

Can you take into account my comment, or discuss it?

**Comment by sergelogvinov** [comment]
I've changed the CSR checker. @PhilippeChepy

File: exoscale/instances.go [hunk 2]
```diff
@@ -93,10 +94,30 @@ func (i *instances) NodeAddressesByProviderID(ctx context.Context, providerID st
...
 
-	return []v1.NodeAddress{{
+	addresses := []v1.NodeAddress{{
 		Type:    v1.NodeExternalIP,
 		Address: instance.PublicIPAddress.String(),
-	}}, nil
+	}}
+
+	if i.p.client != nil && instance.PrivateNetworkIDs != nil && len(*instance.PrivateNetworkIDs) > 0 {
+		if node, _ := i.p.kclient.CoreV1().Nodes().Get(context.TODO(), *instance.Name, metav1.GetOptions{}); node != nil {
```
Comment (PhilippeChepy) [inline exoscale/instances.go:103]: Is there any reason for not using `ctx`?

```suggestion
		if node, _ := i.p.kclient.CoreV1().Nodes().Get(ctx, *instance.Name, metav1.GetOptions{}); node != nil {
```

**Review by PhilippeChepy — COMMENTED** [review]
Integration tests pass with both `kubeadm` and `sks`.

File: exoscale/instances.go (hunk 2 above)
Comment (sergelogvinov) [reply exoscale/instances.go:103]: Oh, sorry, i miss that
I guess this is copy/past issue )

**Review by PhilippeChepy — APPROVED** [review]

**Review by exo-cedric — COMMENTED** [review]
Hi @sergelogvinov 

Thanks for this contribution. Apart from the pending suggestions, looks great!

Note-to-selves: as follow up, we ought to add IPv6-specific tests in [instances_test.go](https://github.com/exoscale/exoscale-cloud-controller-manager/blob/master/exoscale/instances_test.go) and [exoscale_config_test.go](https://github.com/exoscale/exoscale-cloud-controller-manager/blob/master/exoscale/exoscale_config_test.go)

File: exoscale/sks_agent_runner_node_csr_validation.go [hunk 3]
```diff
@@ -104,30 +105,46 @@ func (r *sksAgentRunnerNodeCSRValidation) run(ctx context.Context) {
...
 
+				csrOK := false
 				for _, instance := range instances {
 					if *instance.Name == parsedCSR.DNSNames[0] {
-						if instance.PublicIPAddress.Equal(parsedCSR.IPAddresses[0]) {
-							ok = true
-							break
+						var nodeAddrs, csrAddrs []string
+
+						nodeAddrs = append(csrAddrs, instance.PublicIPAddress.String())
```
[the remark this answers is not in the document]
Comment (sergelogvinov) [reply exoscale/sks_agent_runner_node_csr_validation.go]: yep, artifact from previous refactoring...

instance.PublicIPAddress can be nil too...

[the remark this answers is not in the document]
Comment (sergelogvinov) [reply exoscale/sks_agent_runner_node_csr_validation.go]: i'll add the test little bit later...

**Comment by sergelogvinov** [comment]
@exo-cedric Hi, I've added a few cases to TestNodeAddresses(). And it calls func NodeAddressesByProviderID too. If it make sense I can copy the same code to TestNodeAddressesByProviderID...

I did not find here [exoscale_config_test.go](https://github.com/exoscale/exoscale-cloud-controller-manager/blob/master/exoscale/exoscale_config_test.go) the right place to add tests... 

Can you add more details?

**Review by kobajagi — APPROVED** [review]

**Review by exo-cedric — APPROVED** [review]
We'll deal with the remaining tests ourselves ;-)

Thank you @sergelogvinov for this contribution!

**Pull request merged**